File: src/main/java/me/tigermouthbear/nebulous/Nebulous.java
package me.tigermouthbear.nebulous;

import me.tigermouthbear.nebulous.config.ArrayConfig;
import me.tigermouthbear.nebulous.config.ConfigReader;
import me.tigermouthbear.nebulous.config.StringConfig;
import me.tigermouthbear.nebulous.modifiers.Modifier;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.tree.ClassNode;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.jar.JarOutputStream;

public class Nebulous
{
	private static StringConfig input = new StringConfig("input");
	private static StringConfig output = new StringConfig("output");
	private static ArrayConfig dependencies = new ArrayConfig("dependencies");

	private Map<String, byte[]> files = new HashMap<>();
	private Map<String, ClassNode> classNodes = new HashMap<>();

	public Nebulous(File config)
	{
		ConfigReader.read(config);

		try
		{
			setJar(new JarFile(input.getValue()));
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
	}

	public void apply(Class<? extends Modifier> modifier)
	{
		try
		{
			modifier.newInstance().setTarget(this).modify();
		}
		catch(InstantiationException | IllegalAccessException e)
		{
			e.printStackTrace();
		}
	}

	private void setJar(JarFile jar) {
		Enumeration<JarEntry> entries = jar.entries();
		while (entries.hasMoreElements()) {
			JarEntry entry = entries.nextElement();
			try (InputStream in = jar.getInputStream(entry)) {
				byte[] bytes;
				try (ByteArrayOutputStream tmp = new ByteArrayOutputStream()) {
					byte[] buf = new byte[256];
					for (int n; (n = in.read(buf)) != -1; ) {
						tmp.write(buf, 0, n);
					}
					bytes = tmp.toByteArray();
				}
				if (!entry.getName().endsWith(".class")) {
					files.put(entry.getName(), bytes);
					continue;
				}
				ClassNode c = new ClassNode();
				new ClassReader(bytes).accept(c, ClassReader.EXPAND_FRAMES);
				classNodes.put(c.name, c);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public void saveJar() throws IOException
	{
		String loc = output.getValue();

		if(!loc.endsWith(".jar")) loc += ".jar";

		Path jarPath = Paths.get(loc);
		Files.deleteIfExists(jarPath);
		JarOutputStream outJar = new JarOutputStream(Files.newOutputStream(jarPath, new StandardOpenOption[] {StandardOpenOption.CREATE, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE}));

		//Write classes into obf jar
		for(ClassNode node: getClassNodes().values())
		{
			JarEntry entry = new JarEntry(node.name + ".class");
			outJar.putNextEntry(entry);
			ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_MAXS);
			node.accept(writer);
			outJar.write(writer.toByteArray());
			outJar.closeEntry();
		}

		//Copy files from previous jar into obf jar
		for(Map.Entry<String, byte[]> entry: getFiles().entrySet())
		{
			outJar.putNextEntry(new JarEntry(entry.getKey()));
			outJar.write(entry.getValue());
			outJar.closeEntry();
		}

		outJar.close();
	}

	public Map<String, byte[]> getFiles()
	{
		return files;
	}

	public Map<String, ClassNode> getClassNodes()
	{
		return classNodes;
	}

	public List<String> getDependencies()
	{
		List<String> temp = new ArrayList<>();
		for(int i = 0; i < dependencies.getValue().length(); i++)
			temp.add(dependencies.getValue().getString(i));
		return temp;
	}
}

File: src/main/java/me/tigermouthbear/nebulous/Nebulous.kt
package me.tigermouthbear.nebulous

import me.tigermouthbear.nebulous.config.ArrayConfig
import me.tigermouthbear.nebulous.config.ConfigReader
import me.tigermouthbear.nebulous.config.StringConfig
import me.tigermouthbear.nebulous.modifiers.IModifier
import me.tigermouthbear.nebulous.modifiers.constants.number.NumberPooler
import me.tigermouthbear.nebulous.modifiers.constants.string.StringEncryptor
import me.tigermouthbear.nebulous.modifiers.constants.string.StringPooler
import me.tigermouthbear.nebulous.modifiers.constants.string.StringSplitter
import me.tigermouthbear.nebulous.modifiers.misc.DebugInfoRemover
import me.tigermouthbear.nebulous.modifiers.misc.FullAccessFlags
import me.tigermouthbear.nebulous.modifiers.misc.MemberShuffler
import me.tigermouthbear.nebulous.modifiers.optimizers.GotoInliner
import me.tigermouthbear.nebulous.modifiers.optimizers.GotoReturnInliner
import me.tigermouthbear.nebulous.modifiers.optimizers.LineNumberRemover
import me.tigermouthbear.nebulous.modifiers.optimizers.NOPRemover
import me.tigermouthbear.nebulous.modifiers.renamers.ClassRenamer
import me.tigermouthbear.nebulous.modifiers.renamers.FieldRenamer
import me.tigermouthbear.nebulous.modifiers.renamers.MethodRenamer
import me.tigermouthbear.nebulous.util.ClassPath
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.tree.ClassNode
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarFile
import java.util.jar.JarOutputStream
import java.util.jar.Manifest

/**
 * @author Tigermouthbear
 */

object Nebulous {
	private val input = StringConfig("input")
	private val output = StringConfig("output")
	private val exclusions = ArrayConfig("exclusions")
	private val libraries = ArrayConfig("libraries")

	private val files: MutableMap<String, ByteArray> = HashMap()
	private val classNodes: MutableMap<String, ClassNode> = HashMap()
	private lateinit var manifest: Manifest

	fun run(config: File) {
		val startTime = System.currentTimeMillis()

		ConfigReader.read(config)

		// load jar and libraries
		openJar(JarFile(input.value))
		ClassPath.load(getLibraries())

		val modifiers: List<IModifier> =
		arrayListOf(
				// strings
				StringPooler,
				StringEncryptor,
				StringSplitter,

				// numbers
				NumberPooler,

				// names
				FieldRenamer,
				MethodRenamer,
				ClassRenamer,

				// misc
				MemberShuffler,
				FullAccessFlags,
				DebugInfoRemover,

				// optimizers
				NOPRemover,
				LineNumberRemover,
				GotoInliner,
				GotoReturnInliner
		)

		modifiers.forEach { modifier ->
			val current = System.currentTimeMillis()
			modifier.modify()
			println(modifier.getName() + " completed in " + (System.currentTimeMillis() - current) + " milliseconds")
		}

		saveJar()

		println("\nNebulous finished in " + (System.currentTimeMillis() - startTime) + " milliseconds")
	}

	private fun openJar(jar: JarFile) {
		val entries = jar.entries()

		while(entries.hasMoreElements()) {
			val entry = entries.nextElement()

			jar.getInputStream(entry).use { `in` ->
				val bytes: ByteArray
				val baos = ByteArrayOutputStream()
				val buf = ByteArray(256)
				var n: Int
				while(`in`.read(buf).also { n = it } != -1) {
					baos.write(buf, 0, n)
				}
				bytes = baos.toByteArray()

				if(!entry.name.endsWith(".class")) {
					files[entry.name] = bytes
				} else {
					val c = ClassNode()
					ClassReader(bytes).accept(c, ClassReader.EXPAND_FRAMES)
					classNodes.put(c.name, c)
				}
			}
		}

		// open manifest
		manifest = Manifest(ByteArrayInputStream(files["META-INF/MANIFEST.MF"]))
	}

	private fun saveJar() {
		// save manifest
		val mos = ByteArrayOutputStream()
		manifest.write(mos)
		files["META-INF/MANIFEST.MF"] = mos.toByteArray()

		var location: String = output.value
		if(!location.endsWith(".jar")) location += ".jar"

		val jarPath = Paths.get(location)
		Files.deleteIfExists(jarPath)

		val outJar = JarOutputStream(Files.newOutputStream(jarPath, *arrayOf(StandardOpenOption.CREATE, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)))

        // write classes into jar
		classNodes.values.forEach { cn ->
			outJar.putNextEntry(JarEntry(cn.name + ".class"))

			val writer = ClassWriter(ClassWriter.COMPUTE_MAXS)
			writer.newUTF8("Secured by Nebulous")
			cn.accept(writer)
			outJar.write(writer.toByteArray())

			outJar.closeEntry()
		}

		// copy files into jar
		files.entries.forEach { (key, value) ->
			outJar.putNextEntry(JarEntry(key))
			outJar.write(value)
			outJar.closeEntry()
		}

		outJar.close()
	}

    fun getClassNodes(): MutableMap<String, ClassNode> {
        return classNodes
    }

    fun getFiles(): MutableMap<String, ByteArray> {
        return files
    }

	fun getManifest(): Manifest {
		return manifest
	}

	fun getExclusions(): List<String> {
		val temp: MutableList<String> = mutableListOf()
		for(i in 0 until exclusions.value!!.length()) temp.add(exclusions.value!!.getString(i))
		return temp
	}

	fun getLibraries(): List<String> {
		val temp: MutableList<String> = mutableListOf()
		for(i in 0 until libraries.value!!.length()) temp.add(libraries.value!!.getString(i))
		return temp
	}
}